x-pack/plugins/lens/public/visualizations/xy/persistence.ts

```diff
@@ -0,0 +1,301 @@
+  XYByValueAnnotationLayerConfig,
+} from './types';
+import { isAnnotationsLayer, isByReferenceAnnotationsLayer } from './visualization_helpers';
+import { nonNullable } from '../../utils';
+import { annotationLayerHasUnsavedChanges } from './state_helpers';
+
+export const isPersistedByReferenceAnnotationsLayer = (
+  layer: XYPersistedAnnotationLayerConfig
+): layer is XYPersistedByReferenceAnnotationLayerConfig =>
+  isPersistedAnnotationsLayer(layer) && layer.persistanceType === 'byReference';
+
+export const isPersistedLinkedByValueAnnotationsLayer = (
+  layer: XYPersistedAnnotationLayerConfig
+): layer is XYPersistedLinkedByValueAnnotationLayerConfig =>
+  isPersistedAnnotationsLayer(layer) && layer.persistanceType === 'linked';
+
+/**
+ * This is the type of hybrid layer we get after the user has made a change to
+ * a by-reference annotation layer and saved the visualization without
+ * first saving the changes to the library annotation layer.
+ *
+ * We maintain the link to the library annotation group, but allow the users
+ * changes (persisted in the visualization state) to override the attributes in
+ * the library version until the user
+ * - saves the changes to the library annotation group
+ * - reverts the changes
+ * - unlinks the layer from the library annotation group
+ */
+export type XYPersistedLinkedByValueAnnotationLayerConfig = Omit<
+  XYPersistedByValueAnnotationLayerConfig,
+  'persistanceType'
+> &
+  Omit<XYPersistedByReferenceAnnotationLayerConfig, 'persistanceType'> & {
+    persistanceType: 'linked';
+  };
+
+export type XYPersistedByValueAnnotationLayerConfig = Omit<
+  XYByValueAnnotationLayerConfig,
+  'indexPatternId' | 'hide' | 'simpleView'
+> & { persistanceType?: 'byValue'; hide?: boolean; simpleView?: boolean }; // props made optional for backwards compatibility since this is how the existing saved objects are
+
+export type XYPersistedByReferenceAnnotationLayerConfig = Pick<
+  XYByValueAnnotationLayerConfig,
+  'layerId' | 'layerType'
+> & {
+  persistanceType: 'byReference';
+  annotationGroupRef: string;
+};
+
+export type XYPersistedAnnotationLayerConfig =
+  | XYPersistedByReferenceAnnotationLayerConfig
+  | XYPersistedByValueAnnotationLayerConfig
+  | XYPersistedLinkedByValueAnnotationLayerConfig;
+
+export type XYPersistedLayerConfig =
+  | XYDataLayerConfig
+  | XYReferenceLineLayerConfig
+  | XYPersistedAnnotationLayerConfig;
+
+export type XYPersistedState = Omit<XYState, 'layers'> & {
+  layers: XYPersistedLayerConfig[];
+  valuesInLegend?: boolean;
+  legend: Omit<LegendConfig, 'legendStats'>;
+};
+
+export function convertToRuntime(
+  state: XYPersistedState,
+  annotationGroups?: AnnotationGroups,
+  references?: SavedObjectReference[]
+) {
+  const outputState = needsInjectReferences(state)
+    ? injectReferences(state, annotationGroups, references)
+    : state;
+  if ('valuesInLegend' in outputState) {
+    return convertToLegendStats(outputState);
+  }
+  return outputState;
+}
```

Comment:
Same comment as partition logic, about future proofing this logic for adding more runtime migrations in the future. Also moving the condition check to the migration function.